source/slang/slang-ir-glsl-legalize.cpp

```diff
@@ -864,11 +879,30 @@ GLSLSystemValueInfo* getGLSLSystemValueInfo(
         context->requireGLSLVersion(ProfileVersion::GLSL_450);
         context->requireGLSLExtension(
             UnownedStringSlice::fromLiteral("GL_EXT_fragment_shader_barycentric"));
-        name = "gl_BaryCoordEXT";
 
-        // TODO: There is also the `gl_BaryCoordNoPerspNV` builtin, which
-        // we ought to use if the `noperspective` modifier has been
-        // applied to this varying input.
+        bool hasNoPerspective = false;
+        // Look for interpolation mode decoration in the param chain
+        for (auto paramInfo = outerParamInfo; paramInfo; paramInfo = paramInfo->next)
+        {
+            if (paramInfo->outerParam)
+            {
+                for (auto dd : paramInfo->outerParam->getDecorations())
+                {
+                    if (dd->getOp() == kIROp_InterpolationModeDecoration)
+                    {
+                        auto decoration = (IRInterpolationModeDecoration*)dd;
```

Comment:
Prefer `as<>` for casting. This is a custom function uses for casting, c-style will cause failiures.
Prefer `if(!...) continue;` over nesting `if` statements

```suggestion
                    auto decoration = as<IRInterpolationModeDecoration>(dd);
                    if (!decoration)
                        continue;
```